File: nil/cmd/nil/internal/smartaccount/deploy.go
package smartaccount

import (
	"errors"
	"fmt"

	"github.com/NilFoundation/nil/nil/client/rpc"
	"github.com/NilFoundation/nil/nil/cmd/nil/internal/common"
	"github.com/NilFoundation/nil/nil/internal/types"
	"github.com/NilFoundation/nil/nil/services/cliservice"
	"github.com/NilFoundation/nil/nil/services/cometa"
	"github.com/NilFoundation/nil/nil/services/rpc/jsonrpc"
	"github.com/spf13/cobra"
)

func DeployCommand(cfg *common.Config) *cobra.Command {
	cmd := &cobra.Command{
		Use:   "deploy [path to file] [args...]",
		Short: "Deploy a smart contract",
		Long:  "Deploy the smart contract with the specified hex-bytecode from stdin or from file",
		Args:  cobra.ArbitraryArgs,
		RunE: func(cmd *cobra.Command, args []string) error {
			return runDeploy(cmd, args, cfg)
		},
		SilenceUsage: true,
	}

	setDeployFlags(cmd)

	return cmd
}

func setDeployFlags(cmd *cobra.Command) {
	cmd.Flags().Var(
		types.NewShardId(&params.shardId, types.BaseShardId),
		shardIdFlag,
		"Specify the shard ID to interact with",
	)

	params.salt = *types.NewUint256(0)
	cmd.Flags().Var(
		&params.salt,
		saltFlag,
		"The salt for the deploy transaction",
	)

	cmd.Flags().StringVar(
		&params.AbiPath,
		abiFlag,
		"",
		"The path to the ABI file",
	)

	params.amount = types.Value{}
	cmd.Flags().Var(
		&params.amount,
		amountFlag,
		"The amount of default tokens to send",
	)

	params.token = types.Value{}
	cmd.Flags().Var(
		&params.token,
		"token",
		"The amount of contract token to generate. This operation cannot be performed when the \"no-wait\" flag is set",
	)

	cmd.Flags().BoolVar(
		&params.noWait,
		noWaitFlag,
		false,
		"Define whether the command should wait for the receipt",
	)

	cmd.Flags().StringVar(
		&params.compileInput,
		compileInput,
		"",
		"The path to the JSON file with the compilation input. Contract will be compiled and deployed on the blockchain and the Cometa service",
	)
}

func runDeploy(cmd *cobra.Command, cmdArgs []string, cfg *common.Config) error {
	if !params.token.IsZero() && params.noWait {
		return errors.New("the \"no-wait\" flag cannot be used with the \"token\" flag")
	}

	service := cliservice.NewService(cmd.Context(), common.GetRpcClient(), cfg.PrivateKey, nil)

	var cm *cometa.Client
	if len(params.compileInput) != 0 {
		cm = common.GetCometaRpcClient()
	}

	if len(params.compileInput) == 0 && len(cmdArgs) == 0 {
		return errors.New("at least one arg is required (the path to the bytecode file)")
	}

	var bytecode types.Code
	var err error
	var contractData *cometa.ContractData

	if len(params.compileInput) != 0 {
		contractData, err = cm.CompileContract(params.compileInput)
		if err != nil {
			return fmt.Errorf("failed to compile the contract: %w", err)
		}
		var calldata []byte
		if len(cmdArgs) > 0 {
			abi, err := common.ReadAbiFromFile(params.AbiPath)
			if err != nil {
				return err
			}
			calldata, err = common.ArgsToCalldata(abi, "", cmdArgs)
			if err != nil {
				return fmt.Errorf("failed to pack the constructor arguments: %w", err)
			}
		}
		bytecode = append(contractData.InitCode, calldata...) //nolint:gocritic
	} else {
		var filename string
		var args []string
		if argsCount := len(cmdArgs); argsCount > 0 {
			filename = cmdArgs[0]
			args = cmdArgs[1:]
		}

		bytecode, err = common.ReadBytecode(filename, params.AbiPath, args)
		if err != nil {
			return err
		}
	}

	payload := types.BuildDeployPayload(bytecode, params.salt.Bytes32())

	txnHash, contractAddr, err := service.DeployContractViaSmartAccount(params.shardId, cfg.Address, payload, params.amount)
	if err != nil {
		if errors.Is(err, rpc.ErrTxnDataTooLong) {
			return fmt.Errorf(
				`Failed to marshal transaction: %w.
It appears that your code exceeds the maximum supported size.
Try compiling your contract with the usage of solc --optimize flag,
providing small values to --optimize-runs.
For more information go to
https://ethereum.org/en/developers/tutorials/downsizing-contracts-to-fight-the-contract-size-limit/`, err)
		}
		return err
	}

	var receipt *jsonrpc.RPCReceipt
	if !params.noWait {
		if receipt, err = service.WaitForReceipt(txnHash); err != nil {
			return err
		}
	} else {
		if len(params.compileInput) != 0 {
			return errors.New("the \"no-wait\" flag cannot be used with contract compilation")
		}
	}
	if receipt == nil || !receipt.AllSuccess() {
		return errors.New("deploy transaction processing failed")
	}

	if len(params.compileInput) != 0 {
		if err = cm.RegisterContractData(contractData, contractAddr); err != nil {
			return fmt.Errorf("failed to register the contract: %w", err)
		}
	}

	if !common.Quiet {
		fmt.Print("Transaction hash: ")
	}
	fmt.Printf("0x%x\n", txnHash)

	if !common.Quiet {
		fmt.Print("Contract address: ")
	}
	fmt.Printf("0x%x\n", contractAddr)
	return nil
}

File: clijs/src/commands/smart-account/deploy.ts
import fs from "node:fs";
import path from "node:path";
import type { ContractData, Hex } from "@nilfoundation/niljs";
import { addHexPrefix } from "@nilfoundation/niljs";
import { Args, Flags } from "@oclif/core";
import type { Abi } from "abitype";
import { BaseCommand } from "../../base.js";
import { readJsonFile } from "../../common/utils";
import { bigintFlag } from "../../types";

export default class SmartAccountDeploy extends BaseCommand {
  static override summary = "Deploy a smart contract";
  static override description =
    "Deploy the smart contract with the specified hex-bytecode from stdin or from file";

  static flags = {
    shardId: Flags.integer({
      char: "s",
      description: "Specify the shard ID to interact with",
      required: false,
      default: 1,
    }),
    salt: Flags.integer({
      char: "t",
      description: "The salt for the deploy transaction",
      required: false,
      default: 0,
    }),
    abiPath: Flags.string({
      char: "a",
      description: "The path to the ABI file",
      required: false,
    }),
    amount: bigintFlag({
      char: "m",
      description: "The amount of default tokens to send",
      required: false,
    }),
    token: bigintFlag({
      char: "c",
      description:
        'The amount of contract token to generate. This operation cannot be performed when the "no-wait" flag is set',
      required: false,
      dependsOn: ["tokenName"],
    }),
    tokenName: Flags.string({
      char: "C",
      description: "The name of the token to generate, required when the token flag is set",
      required: false,
      dependsOn: ["token"],
    }),
    noWait: Flags.boolean({
      char: "n",
      description: "Define whether the command should wait for the receipt",
      required: false,
      default: false,
    }),
    compileInput: Flags.string({
      char: "i",
      description:
        "The path to the JSON file with the compilation input. Contract will be compiled and deployed on the blockchain and the Cometa service",
      required: false,
    }),
    fee: bigintFlag({
      char: "f",
      description: "Fee credit for deployment",
      required: false,
    }),
  };

  static args = {
    filename: Args.string({
      name: "filename",
      required: false,
      description: "The path to the bytecode file",
    }),
    args: Args.string({
      name: "args",
      required: false,
      description: "Constructor arguments for the contract",
      multiple: true,
    }),
  };

  static override examples = ["<%= config.bin %> <%= command.id %>"];

  static defaultFee = 100000000000000n;

  public async run(): Promise<Hex> {
    const { flags, args } = await this.parse(SmartAccountDeploy);

    if (flags.noWait) {
      if (flags.token) {
        this.error("The token flag cannot be set when the no-wait flag is set");
      }
      if (flags.compileInput) {
        this.error("The compileInput flag cannot be set when the no-wait flag is set");
      }
    }

    const { smartAccount } = await this.setupSmartAccount();

    let abi: Abi | undefined;

    if (flags.abiPath) {
      try {
        abi = readJsonFile<Abi>(flags.abiPath);
      } catch (e) {
        this.error(`Invalid ABI file: ${e}`);
      }
    }

    let bytecode: Hex | Uint8Array;
    let contractData = {} as ContractData;

    if (flags.compileInput) {
      const cometaClient = this.cometaClient ?? this.error("Cometa client is not initialized");
      const compileInputPath = path.resolve(flags.compileInput);
      const compileInputContent = fs.readFileSync(compileInputPath, "utf8");
      contractData = await cometaClient.compileContract(compileInputContent);
      bytecode = contractData.code;
      abi = contractData.abi as unknown as Abi;
    } else {
      const filename = args.filename;
      if (!filename) {
        this.error("at least one arg is required (the path to the bytecode file");
      }
      const fullPath = path.resolve(filename);
      bytecode = addHexPrefix(fs.readFileSync(fullPath, "utf8"));
    }

    const params = {
      shardId: flags.shardId,
      bytecode: bytecode,
      abi: abi,
      args: args.args?.split(" ") ?? [],
      salt: BigInt(flags.salt),
      value: BigInt(flags.amount ?? 0),
      feeCredit: flags.fee ?? SmartAccountDeploy.defaultFee,
    };
    const { tx, address } = await smartAccount.deployContract(params);

    if (flags.quiet) {
      this.log(address);
    } else {
      this.log("Contract address: ", address);
    }

    if (flags.noWait) {
      return address;
    }

    this.info("Waiting for the contract to be deployed...");
    // await this.waitOnTx(hash);
    await tx.wait();
    this.info("Contract successfully deployed");

    if (flags.compileInput) {
      const cometaClient = this.cometaClient ?? this.error("Cometa client is not initialized");
      await cometaClient.registerContractData(contractData, address);
    }

    if (flags.token) {
      const name =
        flags.tokenName ?? this.error("Token name is required when the token flag is set");

      let tx = await smartAccount.setTokenName(name);
      this.info("Waiting for the token name to be set...");
      await tx.wait();
      this.info("Token name successfully set");

      tx = await smartAccount.mintToken(flags.token);
      this.info("Waiting for the token to be minted...");
      await tx.wait();
      this.info("Token successfully minted");
    }

    return address;
  }
}
